### ISSUE TITLE
Package manager display error while connection succeded

### ISSUE THREAD

**Issue by MikhailArkhipov** [Microsoft/RTVS#2833]
Also, resizing REPL removes the error from PM window. This needs to be more persistent or PM should close per #2809. If you close and reopen window, list is populated.

![image](https://cloud.githubusercontent.com/assets/12820357/20900085/d671a9d4-bae1-11e6-87c8-9d6e05b9c025.png)

**Comment by AlexanderSher** [comment]
Do you have a specific repro? I couldn't get into this state.

**Comment by MikhailArkhipov** [comment]
Occasional repro.

I think it might be b/c there are operations over `_errorMessages` under lock as well as without. Also, there is `Clear` that empties the collection as well as does other things ans `RemoveAllMessages` that only clears and without the lock.

**Comment by AlexanderSher** [comment]
This message, "Connection to R host process failed", can be raised only from `GetPackagesAsync`:
```cs
        private async Task<IReadOnlyList<RPackage>> GetPackagesAsync(Func<Task<JArray>> queryFunc, CancellationToken cancellationToken) {
            // Fetching of installed and available packages is done in a
            // separate package query session to avoid freezing the REPL.
            try {
                var startupInfo = new RHostStartupInfo { Name = "PackageManager" };

                await _session.EnsureHostStartedAsync(startupInfo, null, cancellationToken: cancellationToken);
                await _session.SetVsCranSelectionAsync(CranMirrorList.UrlFromName(_settings.CranMirror), cancellationToken);
                await _session.SetCodePageAsync(_settings.RCodePage, cancellationToken);

                using (await _session.BeginInteractionAsync(true, cancellationToken)) {}

                // Get the repos and libpaths from the REPL session and set them
                // in the package query session
                var repositories = (await DeparseRepositoriesAsync());
                if (repositories != null) {
                    await WrapRException(_session.ExecuteAsync($"options(repos=eval(parse(text={repositories.ToRStringLiteral()})))", cancellationToken));
                }

                var libraries = (await DeparseLibrariesAsync());
                if (libraries != null) { 
                    await WrapRException(_session.ExecuteAsync($".libPaths(eval(parse(text={libraries.ToRStringLiteral()})))", cancellationToken));
                }

                var result = await WrapRException(queryFunc());
                return result.Select(p => p.ToObject<RPackage>()).ToList().AsReadOnly();

            } catch (RHostDisconnectedException ex) {
                throw new RPackageManagerException(Resources.PackageManager_TransportError, ex);
            }
        }
```

At the beginning of the method there is a call `EnsureHostStartedAsync` which can be executed only before or after `RSessionProvider.TrySwitchBrokerAsync`. So if `RHostDisconnectedException` has happened in this method, it either means that 3 seconds (default timeout) isn't enough to connect to the broker (lock waiting time is not included into timeout), or there is some bug that is revealed by this sequence of evaluation requests.

It is also possible that `GetPackagesAsync` has never completed and message has come from request that was canceled during switch. For example, bug that I've fixed in #2830 could cause `session.SetVsCranSelectionAsync` to hang after first call to the `EnsureHostStartedAsync`, while the next time (when session is started) there will be no hanging.

**Comment by MikhailArkhipov** [comment]
Closing, will reopen if I see it again under debugger
